NNC depthwise conv2d implementation

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55213 Integrate NNC conv2d with fuser
* #55025 [nnc] Move device type checks from isSupported to typesAreSupported
* **#54920 NNC depthwise conv2d implementation**

Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

bertmaher committed:
NNC depthwise conv2d implementation

Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]

bertmaher committed:
Update on "NNC depthwise conv2d implementation"


Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_conv.cpp:
```diff
@@ -10,6 +12,44 @@ namespace jit {
+  te::LoopNest loop({output});
+  loop.simplify();
+  loop.prepareForCodegen();
+  te::LLVMCodeGen cg(loop.root_stmt(), {input, weight, bias, output});
```
This might need `ifdef TORCH_ENABLE_LLVM`.

asuhan replied:
Is there any other reason for it except the slowness of the IR evaluator?

ZolotukhinM replied:
It's probably the reason of the linker errors on this PR: if pytorch is built without LLVM, `te::LLVMCodeGen` will be an undefined symbol.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/conv2d.h:
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <torch/csrc/jit/tensorexpr/loopnest.h>
```
Nit: this include could be moved to the cpp file.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/conv2d.cpp:
```diff
@@ -0,0 +1,83 @@
+    BufHandle input,
+    BufHandle weight,
+    BufHandle bias,
+    int stride,
+    int pad,
```
Do we plan to support different W and H strides and pads?

ZolotukhinM replied:
I don't have a strong opinion on this, was just curious :)

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/conv2d.cpp:
```diff
@@ -0,0 +1,83 @@
+  TORCH_INTERNAL_ASSERT(input.ndim() == 4);
+  TORCH_INTERNAL_ASSERT(weight.ndim() == 4);
+
+  auto const& N = immediateAs<int>(input.dim(0));
```
Should we check if dims are actually immediates or are we ok with an assertion in `immediateAs<int>`?

bertmaher replied:
yeah, good point, `immediateAs` will give a really opaque error message like "UNSUPPORTED DTYPE" (which is maybe a separate thing to fix but not quite the fault of this diff :-p ).  I'll add an explicit check.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/conv2d.cpp:
```diff
@@ -0,0 +1,83 @@
+  if (R == 3 && stride == 2 && pad == 1) {
+    For *head, *tail;
+    auto loops = nest.getLoopStmtsFor(conv);
+    nest.sliceHead(loops[kLoopW], 2, &head, &tail);
```
Note for myself: we probably should add a version of `sliceHead` that doesn't require the last two arguments.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/expr.h:
```diff
@@ -10,6 +10,8 @@
 #include <torch/csrc/jit/tensorexpr/mem_arena.h>
 #include <torch/csrc/jit/tensorexpr/types.h>
 
+#include <ATen/core/functional.h>
```
Is it only for `fmap`? You could use `ExprVectorToExprHandleVector` if that doesn't look too ugly :)

bertmaher replied:
Yeah just for fmap.  I have this allergy to super long function names :-p

ZolotukhinM reviewed (approved):
Looks good to me! I have only minor comments.

bertmaher committed:
address comments on "NNC depthwise conv2d implementation"


Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]

bertmaher committed:
TORCH_API on "NNC depthwise conv2d implementation"


Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]

bertmaher committed:
rebase on "NNC depthwise conv2d implementation"


Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]

bertmaher committed:
rebase on "NNC depthwise conv2d implementation"


Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

[ghstack-poisoned]
